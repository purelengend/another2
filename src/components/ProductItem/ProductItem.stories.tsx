import type { Meta, Story } from '@storybook/react'
import { BrowserRouter } from 'react-router-dom'
import ProductItem from './index'
import { ProductsItemProps } from '@constants'

const meta = {
  title: 'Practice Two/ProductItem Component',
  component: ProductItem,
  decorators: [
    Story => (
      <BrowserRouter>
        <div
          style={{
            backgroundColor: '#532EE7',
            padding: '20px',
          }}
        >
          <Story />
        </div>
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} as Meta

export default meta

const Template: Story<{ product: ProductsItemProps }> = args => (
  <ProductItem {...args} />
)

const exampleProduct: ProductsItemProps = {
  id: 'ID01',
  name: 'iPhone 14 Pro Max',
  image: 'https://assets.nicepagecdn.com/d2cc3eaa/3159880/images/yttyy.jpg',
}

export const Primary = Template.bind({})
Primary.args = {
  product: exampleProduct,
}

const exampleProductWithoutImage: ProductsItemProps = {
  id: 'ID01',
  name: 'iPhone 14 Pro Max',
  image: '',
}

export const ProductItemWithoutImage = Template.bind({})
ProductItemWithoutImage.args = {
  product: exampleProductWithoutImage,
}
