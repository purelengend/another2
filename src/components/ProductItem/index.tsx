import React, { memo } from 'react'
import { GridItem, Box, Image, Text, Link } from '@chakra-ui/react'
import imageNotAvailable from '@assets/images/Image_not_available.webp'
import { Link as RouterLink } from 'react-router-dom'
import { IProduct } from '@types'

interface ProductItemProps {
  product: Pick<IProduct, 'id' | 'name' | 'image'>
}

const ProductItem: React.FC<ProductItemProps> = ({ product }) => (
  <GridItem
    className="product-item"
    margin="0 auto"
    key={product.id}
    textAlign="center"
    as="li"
    listStyleType="none"
  >
    <Box
      className="product-item-content"
      marginTop="50px"
      backgroundColor="background"
      borderTop="8px solid"
      borderColor="secondary"
      borderRadius="0 0 20px 20px"
      p={{ base: 2, sm: 4 }}
      boxShadow="5px 5px 29px 0 rgba(0,0,0,0.15)"
      w="full"
      h={{ base: '516px', md: '482px', lg: '318px', xl: '482px' }}
      padding={{ base: '30px 0', sm: '30px 20px' }}
      position="relative"
    >
      <Box className="image-container">
        <Image
          src={product.image || imageNotAvailable}
          alt={product.name}
          w={{
            base: '340px',
            md: '305px',
            lg: '173px',
            xl: '220px',
          }}
          h={{
            base: '300px',
            md: '305px',
            lg: '173px',
            xl: '220px',
          }}
          margin="0 auto"
          objectFit="contain"
          loading="lazy"
        />
      </Box>
      <Text
        fontWeight="semibold"
        padding="15px 0 0"
        fontSize={{ base: '30px', md: '24px', lg: '17px' }}
        color="primary"
        noOfLines={2}
        width={{ xs: '300px', sm: 'full' }}
        margin="0 auto"
        fontFamily="RobotoCondensed-Regular"
      >
        {product.name}
      </Text>
      <Link
        as={RouterLink}
        to={`/products/${product.id}`}
        color="textPrimary"
        position="absolute"
        bottom={{ base: '3%', lg: '3%', xl: '20%' }}
        left="50%"
        transform="translate(-50%, -50%)"
        fontSize="14px"
        fontFamily="RobotoCondensed-Bold"
        textTransform="uppercase"
        opacity="0.5"
        textDecoration="underline"
        _hover={{
          color: 'secondary',
          opacity: '1',
        }}
      >
        detail
      </Link>
    </Box>
  </GridItem>
)

export default memo(ProductItem)
