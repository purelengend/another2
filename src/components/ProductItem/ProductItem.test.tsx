import { render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import ProductItem from './'

jest.mock('@assets/images/Image_not_available.webp', () => ({
  default: 'imageNotAvailable',
}))

describe('ProductItem Component', () => {
  const mockProduct = {
    id: 'P01',
    name: 'Sample Product',
    image: 'sample-image.jpg',
    discount: 11,
    oldPrice: 1263.8,
    newPrice: 1116.31,
    rate: 4.5,
    description: 'Sample description',
    category: 'Phone',
  }

  const mockWithoutImage = {
    id: 'P01',
    name: 'Sample Product',
    image: '',
    discount: 11,
    oldPrice: 1263.8,
    newPrice: 1116.31,
    rate: 4.5,
    description: 'Sample description',
    category: 'Phone',
  }

  it('renders product name', () => {
    const { getByText } = render(
      <MemoryRouter>
        <ProductItem product={mockProduct} />
      </MemoryRouter>
    )

    const productName = getByText('Sample Product')
    expect(productName).toBeInTheDocument()
  })

  it('renders product image', () => {
    const { getByAltText } = render(
      <MemoryRouter>
        <ProductItem product={mockProduct} />
      </MemoryRouter>
    )

    const productImage = getByAltText('Sample Product')
    expect(productImage).toBeInTheDocument()
    expect(productImage).toHaveAttribute('src', 'sample-image.jpg')
  })

  it('renders detail link', () => {
    const { getByText } = render(
      <MemoryRouter>
        <ProductItem product={mockProduct} />
      </MemoryRouter>
    )

    const detailLink = getByText('detail')
    expect(detailLink).toBeInTheDocument()
    expect(detailLink).toHaveAttribute('href', '/products/P01')
  })

  it('matches the snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <ProductItem product={mockProduct} />
      </MemoryRouter>
    )
    expect(container).toMatchSnapshot()
  })

  it('renders imageNotAvailable when product image is not available', () => {
    const { container } = render(
      <MemoryRouter>
        <ProductItem product={mockWithoutImage} />,
      </MemoryRouter>
    )
    expect(container).toMatchSnapshot()
  })
})
