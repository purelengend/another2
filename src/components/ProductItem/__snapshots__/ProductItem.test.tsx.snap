// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem Component matches the snapshot 1`] = `
<div>
  <li
    class="product-item css-gon1ka"
  >
    <div
      class="product-item-content css-8ptibi"
    >
      <div
        class="image-container css-0"
      >
        <img
          alt="Sample Product"
          class="chakra-image css-qzwr75"
          loading="lazy"
          src="sample-image.jpg"
        />
      </div>
      <p
        class="chakra-text css-1lcl2yy"
      >
        Sample Product
      </p>
      <a
        class="chakra-link css-ofgajc"
        href="/products/P01"
      >
        detail
      </a>
    </div>
  </li>
</div>
`;

exports[`ProductItem Component renders imageNotAvailable when product image is not available 1`] = `
<div>
  <li
    class="product-item css-gon1ka"
  >
    <div
      class="product-item-content css-8ptibi"
    >
      <div
        class="image-container css-0"
      >
        <img
          alt="Sample Product"
          class="chakra-image css-qzwr75"
          loading="lazy"
          src="[object Object]"
        />
      </div>
      <p
        class="chakra-text css-1lcl2yy"
      >
        Sample Product
      </p>
      <a
        class="chakra-link css-ofgajc"
        href="/products/P01"
      >
        detail
      </a>
    </div>
  </li>
  ,
</div>
`;
